File: go/internal/bench/runner.go
package bench

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// IndexEntry mirrors the JSON structure of data/index.json.
type IndexEntry struct {
	ID           string `json:"id"`
	Day          string `json:"day"`
	DayPath      string `json:"day_path"`
	VersionCount int    `json:"version_count"`
}

// LoadIndex reads data/index.json.
func LoadIndex(path string) ([]IndexEntry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var idx []IndexEntry
	if err := json.Unmarshal(data, &idx); err != nil {
		return nil, err
	}
	return idx, nil
}

// DayPool builds a map of day -> []IndexEntry for days with >= minEntries entries.
func DayPool(idx []IndexEntry, minEntries int) map[string][]IndexEntry {
	m := make(map[string][]IndexEntry)
	for _, e := range idx {
		m[e.Day] = append(m[e.Day], e)
	}
	out := make(map[string][]IndexEntry)
	for day, es := range m {
		if len(es) >= minEntries {
			out[day] = es
		}
	}
	return out
}

// Op is a single invocation of a benchmark operation.
// Returns (latency, ok). ok=false means this call should be skipped
// (e.g. pool exhausted for create_version); the runner will not count it.
type Op func() (time.Duration, bool)

// RunAdaptive runs op in rounds of batchSize until the 95% CI for the median
// is narrower than precision (relative to the median), or maxN total valid
// samples are collected. batchSize is never decreased below its initial value.
// Returns the collected timings and whether convergence was achieved.
func RunAdaptive(op Op, batchSize int, precision float64, maxN int) ([]time.Duration, bool) {
	timings := make([]time.Duration, 0, batchSize)
	collectBatch := func(n int) {
		consecutiveFails := 0
		for len(timings) < cap(timings) && consecutiveFails < 20 {
			d, ok := op()
			if ok {
				timings = append(timings, d)
				consecutiveFails = 0
			} else {
				consecutiveFails++
			}
		}
		_ = n
	}

	// Use a simpler approach: collect in rounds, check convergence between rounds.
	collected := 0
	for collected < maxN {
		need := min(batchSize, maxN-collected)
		// Grow cap if needed
		if cap(timings)-len(timings) < need {
			next := make([]time.Duration, len(timings), len(timings)+need)
			copy(next, timings)
			timings = next
		}
		_ = collectBatch

		consecutiveFails := 0
		for i := 0; i < need && consecutiveFails < 20; {
			d, ok := op()
			if ok {
				timings = append(timings, d)
				collected++
				consecutiveFails = 0
				i++
			} else {
				consecutiveFails++
			}
		}

		if consecutiveFails >= 20 {
			break // op is exhausted (e.g. pool drained)
		}

		if medianConverged(timings, precision) {
			return timings, true
		}
	}

	return timings, medianConverged(timings, precision)
}

// medianConverged returns true if the 95% CI for the median (computed via
// order statistics) is narrower than precision relative to the median.
// Requires at least 30 samples; returns false below that threshold.
func medianConverged(timings []time.Duration, precision float64) bool {
	n := len(timings)
	if n < 30 {
		return false
	}
	sorted := sortedDurations(timings)
	medIdx := n / 2
	median := sorted[medIdx]
	if median == 0 {
		return true
	}
	// 95% CI for median via order statistics: span ±ceil(0.98*sqrt(n)) from median index.
	// Derived from the binomial: P(X_lo ≤ median ≤ X_hi) ≥ 0.95 where
	// lo = floor(n/2 - 0.98*sqrt(n)), hi = ceil(n/2 + 0.98*sqrt(n)).
	k := int(math.Ceil(0.98 * math.Sqrt(float64(n))))
	lo := medIdx - k
	hi := medIdx + k
	if lo < 0 {
		lo = 0
	}
	if hi >= n {
		hi = n - 1
	}
	relWidth := float64(sorted[hi]-sorted[lo]) / float64(median)
	return relWidth < precision
}

// MedianCIWidth returns the relative 95% CI width for the median of a sorted slice.
// Returns 1.0 if there are fewer than 30 samples.
func MedianCIWidth(sorted []time.Duration) float64 {
	n := len(sorted)
	if n < 30 {
		return 1.0
	}
	medIdx := n / 2
	median := sorted[medIdx]
	if median == 0 {
		return 0.0
	}
	k := int(math.Ceil(0.98 * math.Sqrt(float64(n))))
	lo := medIdx - k
	hi := medIdx + k
	if lo < 0 {
		lo = 0
	}
	if hi >= n {
		hi = n - 1
	}
	return float64(sorted[hi]-sorted[lo]) / float64(median)
}

func sortedDurations(timings []time.Duration) []time.Duration {
	s := make([]time.Duration, len(timings))
	copy(s, timings)
	sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
	return s
}

// Result holds per-operation timing data.
type Result struct {
	Backend   string
	Operation string
	Timings   []time.Duration
	Converged bool
}

func (r *Result) Stats() Stats {
	return ComputeStats(r.Timings)
}

// Stats holds computed latency statistics.
type Stats struct {
	N        int
	Min      time.Duration
	Median   time.Duration
	P95      time.Duration
	P99      time.Duration
	Max      time.Duration
	OpsSec   float64
	MedianCI float64 // relative 95% CI width for the median
}

// ComputeStats computes min/median/p95/p99/max/ops-per-sec from a slice of durations.
func ComputeStats(timings []time.Duration) Stats {
	if len(timings) == 0 {
		return Stats{}
	}
	sorted := sortedDurations(timings)
	n := len(sorted)
	median := sorted[n/2]
	p95 := sorted[int(math.Ceil(float64(n)*0.95))-1]
	p99 := sorted[int(math.Ceil(float64(n)*0.99))-1]
	opsSec := 0.0
	if median > 0 {
		opsSec = float64(time.Second) / float64(median)
	}
	return Stats{
		N:        n,
		Min:      sorted[0],
		Median:   median,
		P95:      p95,
		P99:      p99,
		Max:      sorted[n-1],
		OpsSec:   opsSec,
		MedianCI: MedianCIWidth(sorted),
	}
}

func fmtDur(d time.Duration) string {
	ms := float64(d) / float64(time.Millisecond)
	return fmt.Sprintf("%.2fms", ms)
}

func fmtCI(ci float64, converged bool) string {
	pct := ci * 100
	if !converged {
		return fmt.Sprintf("%4.1f%%!", pct)
	}
	return fmt.Sprintf("%5.1f%%", pct)
}

// PrintTable prints the results in the ASCII table format.
// The medCI column shows the relative 95% CI width for the median;
// a trailing '!' means the precision target was not reached.
func PrintTable(results []*Result, population int) {
	fmt.Printf("Runtime: go  |  Population: %d  |  Date: %s\n\n",
		population, time.Now().Format("2006-01-02 15:04:05"))
	fmt.Printf("%-10s | %-15s | %5s | %6s | %7s | %7s | %7s | %7s | %7s | %7s\n",
		"Backend", "Operation", "N", "medCI", "Min", "Median", "P95", "P99", "Max", "ops/sec")
	fmt.Println(strings.Repeat("-", 100))
	for _, r := range results {
		s := r.Stats()
		fmt.Printf("%-10s | %-15s | %5d | %6s | %7s | %7s | %7s | %7s | %7s | %7.0f\n",
			r.Backend, r.Operation, s.N,
			fmtCI(s.MedianCI, r.Converged),
			fmtDur(s.Min), fmtDur(s.Median), fmtDur(s.P95), fmtDur(s.P99), fmtDur(s.Max),
			s.OpsSec,
		)
	}
}

// SaveCSV writes per-operation nanosecond timings to a CSV file.
func SaveCSV(resultsDir, lang, backend, operation string, timings []time.Duration) error {
	dir := filepath.Join(resultsDir, lang)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	path := filepath.Join(dir, fmt.Sprintf("%s_%s_timings.csv", backend, operation))
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()
	for _, d := range timings {
		fmt.Fprintf(f, "%d\n", d.Nanoseconds())
	}
	return nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: node/runner.js
'use strict';

const fs = require('fs');
const path = require('path');

function loadIndex(dataDir) {
  const data = fs.readFileSync(path.join(dataDir, 'index.json'), 'utf8');
  return JSON.parse(data);
}

function buildDayPool(index, minEntries = 10) {
  const m = new Map();
  for (const e of index) {
    if (!m.has(e.day)) m.set(e.day, []);
    m.get(e.day).push(e);
  }
  const pool = new Map();
  for (const [day, entries] of m) {
    if (entries.length >= minEntries) pool.set(day, entries);
  }
  return pool;
}

// 95% CI for the median via order statistics.
// Returns the relative CI width (CI_width / median).
// Returns 1.0 if fewer than 30 samples.
function medianCIWidth(sorted) {
  const n = sorted.length;
  if (n < 30) return 1.0;
  const medIdx = Math.floor(n / 2);
  const median = sorted[medIdx];
  if (median === 0n) return 0.0;
  const k = Math.ceil(0.98 * Math.sqrt(n));
  const lo = Math.max(0, medIdx - k);
  const hi = Math.min(n - 1, medIdx + k);
  return Number(sorted[hi] - sorted[lo]) / Number(median);
}

function medianConverged(sorted, precision) {
  if (sorted.length < 30) return false;
  return medianCIWidth(sorted) < precision;
}

// Run op in rounds of batchSize until the median CI width is below precision,
// or maxN total valid samples are collected.
// op() returns a BigInt nanosecond latency, or null to signal skip (pool exhausted).
function runAdaptive(op, batchSize, precision, maxN) {
  const timings = [];
  let converged = false;

  while (timings.length < maxN) {
    const need = Math.min(batchSize, maxN - timings.length);
    let consecutiveFails = 0;
    let collected = 0;
    while (collected < need && consecutiveFails < 20) {
      const t = op();
      if (t !== null) {
        timings.push(t);
        consecutiveFails = 0;
        collected++;
      } else {
        consecutiveFails++;
      }
    }

    if (consecutiveFails >= 20) break; // op exhausted

    const sorted = [...timings].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
    if (medianConverged(sorted, precision)) {
      converged = true;
      break;
    }
  }

  if (!converged) {
    const sorted = [...timings].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
    converged = medianConverged(sorted, precision);
  }

  return { timings, converged };
}

function computeStats(timingsNs) {
  if (timingsNs.length === 0) return null;
  const sorted = [...timingsNs].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
  const n = sorted.length;
  const median = sorted[Math.floor(n / 2)];
  const p95 = sorted[Math.ceil(n * 0.95) - 1];
  const p99 = sorted[Math.ceil(n * 0.99) - 1];
  const min = sorted[0];
  const max = sorted[n - 1];
  const medianMs = Number(median) / 1e6;
  const opsSec = medianMs > 0 ? 1000 / medianMs : 0;
  const ciWidth = medianCIWidth(sorted);
  return { n, min, median, p95, p99, max, opsSec, medianMs, ciWidth };
}

function fmtDur(ns) {
  return (Number(ns) / 1e6).toFixed(2) + 'ms';
}

function fmtCI(ciWidth, converged) {
  const pct = (ciWidth * 100).toFixed(1);
  return converged ? `${pct}%` : `${pct}%!`;
}

function printTable(results, population) {
  console.log(`Runtime: node  |  Population: ${population}  |  Date: ${new Date().toISOString().replace('T', ' ').slice(0, 19)}\n`);
  const w = [10, 15, 5, 6, 7, 7, 7, 7, 7, 7];
  const hdr = ['Backend', 'Operation', 'N', 'medCI', 'Min', 'Median', 'P95', 'P99', 'Max', 'ops/sec'];
  console.log(
    hdr[0].padEnd(w[0]) + ' | ' + hdr[1].padEnd(w[1]) + ' | ' +
    hdr[2].padStart(w[2]) + ' | ' + hdr[3].padStart(w[3]) + ' | ' +
    hdr[4].padStart(w[4]) + ' | ' + hdr[5].padStart(w[5]) + ' | ' +
    hdr[6].padStart(w[6]) + ' | ' + hdr[7].padStart(w[7]) + ' | ' +
    hdr[8].padStart(w[8]) + ' | ' + hdr[9].padStart(w[9])
  );
  console.log('-'.repeat(105));
  for (const r of results) {
    const s = computeStats(r.timings);
    if (!s) continue;
    console.log(
      r.backend.padEnd(w[0]) + ' | ' + r.operation.padEnd(w[1]) + ' | ' +
      String(s.n).padStart(w[2]) + ' | ' + fmtCI(s.ciWidth, r.converged).padStart(w[3]) + ' | ' +
      fmtDur(s.min).padStart(w[4]) + ' | ' + fmtDur(s.median).padStart(w[5]) + ' | ' +
      fmtDur(s.p95).padStart(w[6]) + ' | ' + fmtDur(s.p99).padStart(w[7]) + ' | ' +
      fmtDur(s.max).padStart(w[8]) + ' | ' + Math.round(s.opsSec).toString().padStart(w[9])
    );
  }
}

function saveCSV(resultsDir, backend, operation, timingsNs) {
  const dir = path.join(resultsDir, 'node');
  fs.mkdirSync(dir, { recursive: true });
  const filePath = path.join(dir, `${backend}_${operation}_timings.csv`);
  fs.writeFileSync(filePath, timingsNs.map(t => t.toString()).join('\n') + '\n');
}

module.exports = { loadIndex, buildDayPool, runAdaptive, computeStats, printTable, saveCSV };
